librz/arch/isa/hexagon/hexagon_arch.c

```diff
@@ -646,24 +755,101 @@ static HexInsnContainer *hex_to_new_pkt(HexState *state, const HexInsnContainer
-	hex_set_pkt_info(&state->rz_asm, hic, p, 0, false);
-	if (p->last_instr_present) {
-		make_next_packet_valid(state, p);
+	pkt->last_access = rz_time_now();
+	hex_set_pkt_info(&state->rz_asm, hic, pkt, 0, false);
+	if (pkt->last_instr_present) {
+		make_next_packet_valid(state, pkt);
 	}
 	return hic;
 }
 
+#if RZ_BUILD_DEBUG
+static char desc_letter_hic(const HexInsnContainer *hic) {
+	char desc = ' ';
+	if (!hic) {
+		desc = ' ';
+	} else if (hic->is_duplex) {
+		desc = hic->bin.sub[0]->identifier != HEX_INS_INVALID_DECODE ? 'v' : 'i';
+		desc = hic->pkt_info.last_insn ? 'l' : desc;
+	} else {
+		desc = hic->bin.insn->identifier != HEX_INS_INVALID_DECODE ? 'v' : 'i';
+		desc = hic->pkt_info.last_insn ? 'l' : desc;
+	}
+	return desc;
+}
+#endif
+
+static void print_state_pkt(const HexState *state, st32 index, HexBufferAction action, const HexInsnContainer *new_hic) {
+#if RZ_BUILD_DEBUG
+	ut32 oldest = 7;
+	ut32 newest = 0;
+	ut64 min_time = 0xffffffffffffffff;
+	ut64 max_time = 0;
+	for (int i = 0; i < HEXAGON_STATE_PKTS; ++i) {
+		const HexPkt *pkt = &state->pkts[i];
+		if (pkt->last_access < min_time) {
+			min_time = pkt->last_access;
+			oldest = i;
+		}
+		if (pkt->last_access > max_time) {
+			max_time = pkt->last_access;
+			newest = i;
+		}
+	}
+	RZ_LOG_DEBUG("╭─────┬──────────────┬─────┬──────────────────┬───────────────╮\n");
+	RZ_LOG_DEBUG("│ pkt │   packet     │     │                  │ [i]n[v]alid   │\n");
+	RZ_LOG_DEBUG("│ id  │   address    │ age │    last access   │ [l]ast        │\n");
+	RZ_LOG_DEBUG("├─────┼──────────────┼─────┼──────────────────┼───┬───┬───┬───┤\n");
+	RzStrBuf *pkt_line = rz_strbuf_new("");
+	for (int i = 0; i < HEXAGON_STATE_PKTS; ++i) {
+		const HexPkt *pkt = &state->pkts[i];
+		const char *time_ind = "   ";
+		if (i == oldest) {
+			time_ind = "old";
+		} else if (i == newest) {
+			time_ind = "new";
+		}
+		rz_strbuf_appendf(pkt_line, "│  %d  │ 0x%010x │ %s │ %016llu │ ", i, pkt->pkt_addr, time_ind, pkt->last_access);
+		HexInsnContainer *hic = NULL;
+		for (int j = 0; j < 4; ++j) {
+			hic = rz_list_get_n(pkt->bin, j);
+			const char desc = desc_letter_hic(hic);
+			rz_strbuf_appendf(pkt_line, "%c │ ", desc);
+		}
+		if ((index < 0 && i == oldest) || (index == i)) {
+			const char desc = desc_letter_hic(new_hic);
+			rz_strbuf_append(pkt_line, " < ");
+			if (action == HEX_BUF_ADD) {
+				rz_strbuf_appendf(pkt_line, "%s %c", "ADDED", desc);
+			} else if (action == HEX_BUF_STALE) {
+				rz_strbuf_appendf(pkt_line, "added %c %s", desc, "to STALE");
+			} else if (action == HEX_BUF_NEW) {
+				rz_strbuf_appendf(pkt_line, "added %c %s", desc, "to NEW");
+			}
+		}
+		rz_strbuf_append(pkt_line, "\n");
+		RZ_LOG_DEBUG("%s", rz_strbuf_get(pkt_line));
+		rz_strbuf_fini(pkt_line);
+		if (i < HEXAGON_STATE_PKTS - 1) {
+			RZ_LOG_DEBUG("├─────┼──────────────┼─────┼──────────────────┼───┼───┼───┼───┤\n");
+		} else {
+			RZ_LOG_DEBUG("╰─────┴──────────────┴─────┴──────────────────┴───┴───┴───┴───╯\n");
+		}
```

Comment:
when `RZ_LOG_DEBUG` id not enabled this is kinda useless. maybe change this as
```c
RZ_LOG_DEBUG(i < HEXAGON_STATE_PKTS - 1 ?
"├─────┼──────────────┼─────┼──────────────────┼───┼───┼───┼───┤\n" :
"╰─────┴──────────────┴─────┴──────────────────┴───┴───┴───┴───╯\n"
);
```

Reply:
Yeah, right. Let me add an early return to the function.

This logging is mostly there, because it visualizes nicely how the packets are filled over time. A thing I found really difficult to track just by looking at the debugger variables. This helps with finding some weird instruction buffering issues.

We can also remove it, if you feel it is unnecessary overhead.